htlcswitch/link_test.go

```diff
@@ -6659,6 +6659,164 @@ func TestShutdownIfChannelClean(t *testing.T) {
 	}
 }
 
+// TestPipelineSettle tests that a link should only pipeline a settle if the
+// related add is fully locked-in meaning it is on both sides' commitment txns.
+func TestPipelineSettle(t *testing.T) {
```

Comment:
Excellent test case reproduction!

Reply:
Would be cool to eventually translate these into some flat file format which can be used on the spec level to test implementations. IIRC we had a very early version of something like this, but it was never fully maintained.

Reply:
Yes i think the spec could benefit from an exhaustive list of state transition diagrams (add, settle, fee, closing signed etc) that show valid and invalid sequences